volatility3/framework/plugins/linux/psscan.py

```diff
@@ -0,0 +1,127 @@
+        Returns:
+            A tuple with the fields to show in the plugin output.
+        """
+        pid = task.tgid
+        tid = task.pid
+        ppid = task.parent.tgid if task.parent else 0
+        name = utility.array_to_string(task.comm)
+
+        task_fields = (format_hints.Hex(task.vol.offset), pid, tid, ppid, name)
+        return task_fields
+
+    def _generator(self):
+        """Generates the tasks found from scanning."""
+
+        for task in self.scan_tasks(
+            self.context, self.config["kernel"], self.config["kernel.symbol_table_name"]
+        ):
+            row = self._get_task_fields(task)
+            yield (0, row)
+
+    @classmethod
+    def scan_tasks(
+        cls,
+        context: interfaces.context.ContextInterface,
+        vmlinux_module_name: str,
+        kernel_table_name: str
+    ) -> Iterable[interfaces.objects.ObjectInterface]:
+        """Scans for tasks in the memory layer.
+
+        Args:
+            context: The context to retrieve required elements (layers, symbol tables) from
+            vmlinux_module_name: The name of the kernel module on which to operate
+        Yields:
+            Task objects
+        """
+        vmlinux = context.modules[vmlinux_module_name]
+
+        # check if this image is 32bit or 64bit
+        is_32bit = not symbols.symbol_table_is_64bit(
+            context, vmlinux.symbol_table_name
+        )
+        if is_32bit:
+            pack_format = "I"
+            kernel_add = 0
+        else:
+            pack_format = "Q"
+            kernel_add = 0xffff000000000000
+
+        # get task_struct to find the offset to the sched_class pointer
+        sched_class_offset = vmlinux.get_type("task_struct").members["sched_class"][0]
+
+        # find all sched_class names by searching by if they include '_sched_class', e.g. 'fair_sched_class'
+        all_sched_class_names = [ symbol for symbol in vmlinux.symbols if '_sched_class' in symbol ]
+
+        # find where all these sched_class would be in this image
+        all_sched_class_addrs = [ vmlinux.get_symbol(sched_class_name).address for sched_class_name in all_sched_class_names ]
+
+        # convert all sched_class addrs into bytes to act as needles for scanning
+        needles = [ struct.pack(pack_format, addr + kernel_add) for addr in all_sched_class_addrs ]
+
+        # scan the memory_layer for these needles
+        memory_layer = context.layers['memory_layer']
+        for address, _ in memory_layer.scan(
+                context,
+                scanners.MultiStringScanner(needles)
+                ):
+            # create task in the memory_layer
+            ptask = context.object(kernel_table_name + constants.BANG + "task_struct", 
+                                    offset=address-sched_class_offset, 
+                                    layer_name='memory_layer')
+            
+            # sanity check exit_state
+            if ptask.exit_state not in [0, 16, 32, 16|32]:
```

Comment:
It would be good to document what these flags are for and/or turn them into symbols or at least an enum, so the names can be referenced and people understand which exit_states are being filtered).

Reply:
Yes, that's a great idea. I've included it in the output too now.